src/main/java/com/deviceinsight/kafka/health/KafkaConsumingHealthIndicator.java

```diff
@@ -73,21 +77,27 @@ public KafkaConsumingHealthIndicator(KafkaHealthProperties kafkaHealthProperties
 		this.consumer = new KafkaConsumer<>(kafkaConsumerPropertiesCopy, deserializer, deserializer);
 		this.producer = new KafkaProducer<>(kafkaProducerProperties, serializer, serializer);
 
-		this.executor = new ThreadPoolExecutor(0, 1, 0L, MILLISECONDS, new SynchronousQueue<>(),
-				new ThreadPoolExecutor.AbortPolicy());
+		this.executor = Executors.newSingleThreadExecutor();
+		this.running = new AtomicBoolean(true);
+		this.kafkaCommunicationResult = KafkaCommunicationResult.failure(topic, null);
 	}
 
 	@PostConstruct
+	@VisibleForTesting
 	void subscribeAndSendMessage() throws InterruptedException {
 		subscribeToTopic();
-		KafkaCommunicationResult kafkaCommunicationResult = sendAndReceiveMessage();
-		if (kafkaCommunicationResult.isFailure()) {
-			throw new RuntimeException("Kafka health check failed", kafkaCommunicationResult.getException());
-		}
+
+		executor.submit(() -> {
+			while (running.get()) {
+				kafkaCommunicationResult = sendAndReceiveMessage();
```

Comment:
I'm unsure about both writing and reading in a loop. I worry that this might put too much load on kafka as we would write at very high frequency.

The alternative would be to only poll in the separate thread and write to kafka directly from the health indicator invocation. We would need to use some concurrent data structure for this. Maybe a caffeine cache would be a good fit as it could automatically expire the entries.
The expiration time would be >= the timeout of the writing thread's wait time for the poll to succeed.